### DISCUSSION TITLE
Encrypt template?

### DISCUSSION THREAD

**Discussion by mauvehed** [twpayne/chezmoi#3713] [Q&A] [answered]
Is it possible to encrypt a .tmpl file or a template stored in .chezmoitemplates ? I'm not seeing a way to do this.

What I'm trying to accomplish, perhaps there's a better way, is to encrypt part of a file and not the whole thing. The file itself is a .tmpl and has a dynamic include {{ template "file" }} that's based on the OS chezmoi is being installed on. It's just that secondary template that I want encrypted.

**Comment by twpayne** [comment, marked as answer]
The easy way to do this is to use the [`decrypt` template function](https://www.chezmoi.io/reference/templates/functions/decrypt/), something like:

```
{{ join .chezmoi.sourceDir ".encrypted-data.txt" | include | decrypt }}
```

Where `.encrypted-data.txt` is the file containing the encrypted data. The leading `.` makes chezmoi ignore it. You can create it with:

```console
$ chezmoi encrypt -o $(chezmoi source-path)/.encrypted-data.txt unencrypted-data.txt
```

**Comment by mauvehed** [comment]
Much appreciated, I have the basics working. But have a few follow up questions.

1. I can _cz edit_ the parent template file, but that won't have any impact on the encrypted file that's being included via template.
2. Is the only way to edit the encrypted file is to go through the manual process of encrypt/decrypt?
3. Can I store the encryption password in 1password and use {{ onepasswordRead }} to some how simplify that process?

(P.S. This project is amazing and I thank you greatly for its creation! I've read your other post on the challenges of OSS and I humbly give you my deepest gratitude for all that you do here)

**Comment by mauvehed** [comment]
You know what.. I think your own config may have given me a better answer. 

https://github.com/twpayne/dotfiles/blob/master/home/private_dot_ssh/config.tmpl

```{{ onepasswordRead "op://FLARM/SSH Config/notesPlain" .onepasswordAccount }}```

This is basically what I was trying to achieve. But I've simplified a few different things, and now realizing I could just load this via 1P. 

I'll still refer to my questions above. As I may need to encrypt some other things in the future. The problem even with 1P is that I can't easily _cz edit_ this stored config, right?

**Reply by halostatue** [reply]
Correct. By and large, file encryption (including the use of `onepasswordDocument`) is recommended when you have easily managed *chunks* of code (I use that for `~/.aws/credentials` chunks in 1Password) or entire files that should be stored on the server. Otherwise, it is recommended to use data retrieval from something like 1Password to get fields to fill in as appropriate.

**Comment by mauvehed** [comment]
Thanks again, @twpayne! Chezmoi has been a life saver I've been looking for the last 25 years!

**Comment by rafaelpirolla** [comment]
Sorry to revive this one @twpayne , but would there be any way to process the template when including it?
Like if in .chezmoi.yaml.tmpl:
```
encryption: "age"
age:
  identity: "~/chezmoi_test/age-key.txt"
  recipient: "age14qnf7407uvqx6sv5z4pcm65qktyrada9f9qlfffz0h60hfqv933q5efp7e"
data:
  test: aleluia
{{ joinPath .chezmoi.sourceDir ".chezmoitemplates/init.data.enc" | includeTemplate | decrypt | indent 2 }}
```
But the decrypted init.data would be:
```
string_test: "example"
pw_example: '{{ onepasswordRead "op://FLARM/SSH Config/notesPlain" }}'
```
I tried but with decrypt it does include the file but it doesn't process the template functions.

**Reply by twpayne** [reply]
What you're proposing here has a circular dependency: the config file contains the private key needed to decrypt the template, but the config file is itself a template that needs the private key before it can be executed.

I suggest moving the encryption to the target files (i.e. where the decrypted values are needed).

**Reply by rafaelpirolla** [reply]
I would like to put my chezmoi directory into a public repo on Github but I would like to hide which password manager I'm using.

Is there any way to decrypt a template before processing it? For example:
.chezmoitemplates/cntlm.enc (decrypted):
```
{{ onepasswordRead "op://FLARM/SSH Config/notesPlain" }}
```
Then:
dot_config/cntlm/cntlm.conf
```
{{ includeTemplate cntlm.enc | decrypt }}
```

Extended the discussion here: https://github.com/twpayne/chezmoi/discussions/4350
But, in short, there's no way to achieve this. Only with the full template being encrypted.
